core/lib/object_store/src/gcs_unauthenticated.rs

```diff
@@ -0,0 +1,99 @@
+
+#[derive(Debug)]
+pub struct UnauthenticatedGoogleCloudStorage {
+    bucket_prefix: String,
+    max_retries: u16,
+}
+
+impl UnauthenticatedGoogleCloudStorage {
+    pub fn new(bucket_prefix: String, max_retries: u16) -> UnauthenticatedGoogleCloudStorage {
+        Self {
+            bucket_prefix,
+            max_retries,
+        }
+    }
+}
+
+#[async_trait]
+impl ObjectStore for UnauthenticatedGoogleCloudStorage {
+    async fn get_raw(&self, bucket: Bucket, key: &str) -> Result<Vec<u8>, ObjectStoreError> {
+        const DOWNLOAD_TIMEOUT: Duration = Duration::from_secs(600);
+        let client = reqwest::Client::builder()
+            .timeout(DOWNLOAD_TIMEOUT)
+            .build()
+            .map_err(|error| ObjectStoreError::Other(error.into()))?;
+
+        let mut last_error: Option<Error> = None;
+        let retry_count = self.max_retries;
+        for retry_number in 0..retry_count {
+            let url = format!("{}/{key}", self.storage_prefix_raw(bucket));
+            let response = client.get(url.to_string()).send().await;
+
+            match response {
+                Ok(response) => {
+                    let response_code = response.status();
+                    match response_code {
+                        StatusCode::NOT_FOUND => {
+                            let error_message = format!(
+                                "missing key: {key} in bucket {bucket} (got 404 from {url})"
+                            );
+                            return Err(ObjectStoreError::KeyNotFound(error_message.into()));
+                        }
+                        StatusCode::OK => {}
+                        _ => {
+                            let error_message = format!("unexpected error when fetching {key} from bucket {bucket}, received code {response_code}");
+                            return Err(ObjectStoreError::Other(error_message.into()));
+                        }
+                    }
+                    let bytes = response.bytes().await.map(|bytes| bytes.to_vec());
+                    match bytes {
+                        Ok(bytes) => return Ok(bytes),
+                        Err(error) => {
+                            last_error = Some(error);
+                        }
+                    }
+                }
+                Err(error) => {
+                    last_error = Some(error);
+                }
+            }
+
+            tracing::warn!(
+                "Failed to download {url} (attempt {}/{retry_count}). Backing off for 5 second",
+                retry_number + 1
+            );
+            tokio::time::sleep(Duration::from_secs(5)).await;
+        }
+        Err(ObjectStoreError::Other(BoxedError::from(
+            last_error.unwrap(),
+        )))
+    }
+
+    async fn put_raw(
+        &self,
+        _bucket: Bucket,
+        _key: &str,
+        _value: Vec<u8>,
+    ) -> Result<(), ObjectStoreError> {
+        unimplemented!("This store is read-only!")
```

Comment:
Bikeshedding: I'd return an error from this method and `remove_raw` since we cannot guarantee at compile-time that these methods won't be called, and dealing with panics is more costly / difficult than dealing with errors. Ideally, it could make sense to move `get_raw` to a new supertrait for `ObjectStore` (`ReadonlyObjectStore`?) and use this trait in snapshot recovery logic, so that it *can* be guaranteed in compile-time. But the downside of this is that we won't be able to configure such a store using `ObjectStoreFactory` (or at least it'd require more effort), so I'm not sure it's worth the trouble.

Reply:
I see now, that it may lead to runtime error, I don't think it's worth it to enforce this on type level as probably for a long time this could be the only read-only implementation. I have one idea though, I can put ReadOnly into the name of the type: PublicReadOnlyGoogleCloudStorage, it should make it way more difficult to miss

Reply:
IMO, it's not really necessary given how public buckets work in GCS, S3 etc.; they imply public readonly access, not full read-write permissions. But I'm OK with this long name as well 🙃